module/userExtension/helper.js

```diff
@@ -211,304 +212,373 @@ module.exports = class UserExtensionHelper {
-                  existingUser.platformRoles[userPlatformRoleToUpdate].programs =
-                    existingUser.platformRoles[userPlatformRoleToUpdate].programs.concat(programIds);
-                  existingUser.platformRoles[userPlatformRoleToUpdate].programs = _.uniq(
-                    existingUser.platformRoles[userPlatformRoleToUpdate].programs,
+                  const currentRoleInfoIndex = existingUserPlatformRoles.findIndex(
+                    (pr) => pr.programId.toString() === programObjectId.toString()
                   );
-                } else if (userRole.programOperation == 'REMOVE') {
-                  if (programIds.length > 0) {
-                    programIds.forEach((programId) => {
-                      _.pull(existingUser.platformRoles[userPlatformRoleToUpdate].programs, programId.toString());
-                    });
+
+                  const newRoles = platform_role_array;
+
+                  if (userRole.programOperation === 'OVERRIDE') {
+                    if (currentRoleInfoIndex !== -1) {
+                      const currentRoles = existingUserPlatformRoles[currentRoleInfoIndex].roles;
+
+                      // Find roles to remove (exist in current but not in new)
+                      const rolesToRemove = currentRoles.filter((role) => !newRoles.includes(role));
+                      for (const role of rolesToRemove) {
+                        kafkaEventPayloads.push({
+                          userId: userProfile.id,
+                          username: userProfile.username,
+                          programId: programObjectId,
+                          role,
+                          eventType: 'delete',
+                        });
+                      }
+
+                      // Find roles to add (exist in new but not in current)
+                      const rolesToAdd = newRoles.filter((role) => !currentRoles.includes(role));
+                      for (const role of rolesToAdd) {
+                        kafkaEventPayloads.push({
+                          userId: userProfile.id,
+                          username: userProfile.username,
+                          programId: programObjectId,
+                          role,
+                          eventType: 'create',
+                        });
+                      }
+
+                      // Override the roles
+                      existingUserPlatformRoles[currentRoleInfoIndex].roles = [...newRoles];
+                    } else {
+                      // Create new program entry
+                      existingUserPlatformRoles.push({
+                        programId: programObjectId,
+                        roles: [...newRoles],
+                      });
+
+                      // All roles are new, emit create events
+                      for (const role of newRoles) {
+                        kafkaEventPayloads.push({
```

Comment:
@borkarsaish65 make sure the event produced is identical to suggested one

{
    "entity": "program",
    "eventType": "create/update/delete",
    "username": "vivek1998",
    "email": "vivek.m@shikshalokam.org",
    "userId": 101,
    "meta": {
      "programInformation": {
        "name": "Program Name",
        "externalId": "PROGRAM_EXT",
        "id": "67c82d7abad58c889bc5a627"
      }
    }
  }
  
  Ignore the comment if you think we are producing events with all the data required. but just replay this comments saying so before you resolve this

Reply:
except email all fields are present , Vivek told he will map the user using username itself